fix: cert rendering temporary

Sembauke opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #42893

<!-- Feel free to add any additional description of changes below this line -->

Sembauke committed:
fix: require login as step

Sembauke committed:
fix: cert rendering temporary

Sembauke committed:
fix: remove keys from other translation files

Sembauke committed:
fix: cypress test

Sembauke commented:
Let me know what you think about the extra step for claiming a certification.

ShaunSHamilton reviewed (dismissed):
Thanks, for tackling this, Sem.

I would prefer not bundling state like `isSignedIn` with the `claimCertSteps`. Would you mind using the selector in the component dispatch instead, and passing it through as a prop?

Sembauke commented:
> Thanks, for tackling this, Sem.
> 
> I would prefer not bundling state like `isSignedIn` with the `claimCertSteps`. Would you mind using the selector in the component dispatch instead, and passing it through as a prop?

This would mean that `isSignedIn` will not count as an extra step. So instead of `0/6`  it would be `0/5`

ShaunSHamilton commented:
> > Thanks, for tackling this, Sem.
> > I would prefer not bundling state like `isSignedIn` with the `claimCertSteps`. Would you mind using the selector in the component dispatch instead, and passing it through as a prop?
> 
> This would mean that `isSignedIn` will not count as an extra step. So instead of `0/6` it would be `0/5`

Correct. Campers who are not signed in should not see the steps card, as per the linked issue.

Sembauke committed:
chore: shauns suggestions

Sembauke committed:
fix: remove steps

Ewelin36 commented:
Dne ne 29. 8. 2021 13:05 uživatel github-actions[bot] <
***@***.***> napsal:

> Thanks for your pull-request.
>
> We are no longer accepting changes to the non-English versions of files in
> parts of this codebase. This pull-request seems to change some of those.
> Please visit our contributing guidelines
> <https://contribute.freecodecamp.org> to learn more about translating
> freeCodeCamp's resources.
>
> As always, we value all of your contributions.
>
> Happy contributing!
> ------------------------------
>
> *Note: This message was automatically generated by a bot. If you feel this
> message is in error or would like help resolving it, feel free to reach us
> in our contributor chat
> <https://chat.freecodecamp.org/channel/contributors>.*
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/freeCodeCamp/freeCodeCamp/pull/43335#issuecomment-907772365>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AUUVTMUJMLW3J2TX6WDYXGTT7IH7JANCNFSM5DAD2JPA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

ShaunSHamilton commented on client/i18n/locales/english/translations.json:
```diff
@@ -567,6 +567,7 @@
     "title": "Claim Your Certification",
     "intro": "Complete the following steps to claim and view your {{i18nCertText}}",
     "complete-project": "Complete {{i18nCertText}} Projects",
+    "login": "Login to claim certification",
```
If you could revert this change

ShaunSHamilton commented on client/src/templates/Introduction/components/CertChallenge.js:
```diff
@@ -82,6 +102,10 @@ const CertChallenge = ({
   const { certSlug } = certMap.find(x => x.title === title);
 
   useEffect(() => {
+    if (complete && !pending) {
+      setUserLoaded(true);
+    }
```
Take this with a grain of salt:

As `complete` and `pending` are props, we should be adding them to the dependency array of **their own** `useEffect`. Also, I do not think they should be _accessed_ (destructured) within the root of the function - should be destructured within the `useEffect` callback.

Sembauke replied:
Thanks for the review 👍

ShaunSHamilton reviewed (changes requested):
Once this is changed, all LGTM. Thanks, Sem.

Sembauke committed:
fix: use a seperate UseEffect function

ShaunSHamilton commented on client/src/templates/Introduction/components/CertChallenge.js:
```diff
@@ -118,10 +145,9 @@ const CertChallenge = ({
   return (
     <div className='block'>
-      {(!isCertified || !canViewCert) && (
+      {!isCertified && userLoaded && isSignedIn && (
```
Correct me, if I am wrong, but this still needs to hold for if a Camper has not accepted the academic honesty etc.
```suggestion
      {(!isCertified || !canViewCert) && userLoaded && isSignedIn && (
```

Sembauke replied:
I hardly could tell the difference between isCertified and canViewCert

ShaunSHamilton replied:
`isCertified` is when a Camper has claimed a certificate. `canViewCert` is for when a Camper has made the correct settings public to allow a certificate to be viewed.

These two could possibly be combined into one, but at some point we might want slightly different behaviour.

Sembauke committed:
Suggestions from Shaun

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

Sembauke commented:
I don't know why but the cert temp renders again because of this change.

Sembauke commented:
So I figured out what is going wrong:

The certification card renders on these conditions:
`userLoaded && isSignedIn && (!isCertified || !canViewCert)`

if a user is not loaded and is certified the following would happen:
` userLoaded: false, isSignedIn: true, isCertified: false,  canViewCert: false)` 

the condition would come out as false.

But when the user is loaded:
 ` userLoaded: true, isSignedIn: true, isCertified: true,  canViewCert: false)` 

the condition would come out as true, thus still temporarily rendering the certification.
This probably happens because the function that counts the steps is not done counting.

In the next render: 
` userLoaded: true, isSignedIn: true, isCertified: true,  canViewCert: true)` 

the condition would come out as false.

-----
I like to hear your thoughts.

ojeytonwilliams commented:
@Sembauke could you expand a bit about what's going wrong?

Sembauke commented:
@ojeytonwilliams 
At the moment the certification steps render for split seconds and then disappear again.  
The explanation above is what I found out.

If you want to take a look yourself go ahead.

Here a slow-motion video:
https://imgur.com/a/3EgXpgT

ojeytonwilliams commented:
Thanks, Sem, I'm with you now.  How about we simply hide the Claim Your Certification component behind a skeleton until the user has loaded?  

It's a bit out of scope for this PR, but if they're not signed in at all, I think it makes sense to have that section be 

> Claim Your Certification
> Sign in to save your progress and claim certifications

Sembauke commented:
> Thanks, Sem, I'm with you now. How about we simply hide the Claim Your Certification component behind a skeleton until the user has loaded?

I do not think that is possible with the currently implemented logic.

That is because the `canViewCert` is not relying on `isSigned` (it is not part of the steps that are asynchronously received).

A solution might be to set `canViewCert` to true if `isCertified` is true, thus not rendering the certification card.

I hope this makes sense.

ojeytonwilliams commented:
I'm probably missing something, but my understanding is that the client can't know exactly what to render until the user object has returned from the api.  Is that correct?

If it is, we should change the render logic so that it shows a skeleton until the user has returned.

Sembauke committed:
fix: certification temp rendering

Sembauke commented:
@ojeytonwilliams I am not sure but I think I fixed it, can you check and see if you like the new behavior?

ShaunSHamilton commented:
This is confusing. I am getting that behaviour again where clicking close on the Donation Modal causes the viewport to jump to the top of the page.

ShaunSHamilton commented on client/src/templates/Introduction/components/CertChallenge.js:
```diff
@@ -118,10 +144,9 @@ const CertChallenge = ({
   return (
     <div className='block'>
-      {(!isCertified || !canViewCert) && (
+      {userLoaded && isSignedIn && !isCertified && !canViewCert && (
```
```suggestion
      {userLoaded && isSignedIn && (!isCertified || !canViewCert) && (
```

ShaunSHamilton replied:
This was silly - I did not read above conversation.

Sembauke replied:
No problem Shaun!

ShaunSHamilton reviewed (changes requested):
I **_THINK_** this is what we need

Sembauke committed:
fix: cert temp rendering hopefully

ShaunSHamilton reviewed (dismissed):
This LGTM 👍 . Nice job, Sem.

ojeytonwilliams commented:
@Sembauke before we get this in, could you update the tests, since we know they don't catch the modal bug?

Sembauke committed:
fix: button visibility and cert test

ojeytonwilliams commented:
Thanks for updating the tests, Sem.  Just so I can QA this properly, which commit had the bug where it jumps to the top of the page?

Sembauke committed:
fix: condition

ojeytonwilliams committed:
test: improve claim cert tests

Without npm run seed, retries will be in a different state.

Check the scroll position to (hopefully) catch the weird cert claim bug.

ShaunSHamilton reviewed (dismissed):
This makes sense with the button not showing to signed-out users. Thanks, Sem.

ojeytonwilliams commented on client/src/templates/Introduction/components/CertChallenge.js:
```diff
@@ -53,10 +69,14 @@ const CertChallenge = ({
   const [canClaim, setCanClaim] = useState({ status: false, result: '' });
   const [isCertified, setIsCertified] = useState(false);
+  const [userLoaded, setUserLoaded] = useState(false);
+  const [userChecked, setUserChecked] = useState(false);
```
What does `userChecked` mean?  I think it would make more sense to have a `verificationFetchState` object to track `getVerifyCanClaimCert`.  Also, if you're feeling keen, `canClaim` really should be a boolean called `certIsClaimable` and `result` should be separate and called something meaningful. `verificationStatus`, maybe.  It's unclear to me what it's for.

@ShaunSHamilton might have more context.

Sembauke replied:
Not sure what you mean by :

>  I think it would make more sense to have a verificationFetchState object to track getVerifyCanClaimCert

ojeytonwilliams replied:
So, we generally use an object

```
{   
  pending: boolean,
  complete: boolean,
  errored: boolean,
  error: null | string
}
```
to track the status of an async operation, so I was suggesting we do something similar here.  Essentially we perform an ajax operation and we conditionally display stuff on the page depending on what state that operation is in.

On reflection, it's probably over the top to use the full object if we only care about the `complete` part of the process.

Anyway, that's a lot of words and not a very concrete recommendation!  So, I'll suggest a code change in another comment.

ojeytonwilliams replied:
So, I figured it was easier to show than tell, and I pushed a commit.

The rationale for the name changes is pretty straightforward.  I don't know what `canClaim.status` or `canClaim.result` mean, but I can easily guess about `canClaimCert` and `certVerificationMessage`.  Well, I still don't know what the latter means, but it *is* a message about certification verification :man_shrugging: 

Same point for `userChecked` vs `verificationComplete`.  The former is a bit vague (what are we checking the user _for_). The latter is quite explicit and mirrors the patterns we usually use for ajax requests.  (pending, complete, etc. etc.).

ojeytonwilliams commented on client/src/templates/Introduction/components/CertChallenge.js:
```diff
@@ -118,27 +146,35 @@ const CertChallenge = ({
-      >
-        {isCertified ? t('buttons.show-cert') : t('buttons.claim-cert')}
-      </Button>
+      {userLoaded &&
+        isSignedIn &&
+        (!isCertified || (!canViewCert && userChecked)) && (
```
```suggestion
      {userLoaded &&
        isSignedIn &&
        (!isCertified || (!canViewCert && userChecked)) && (
```
This is a bit complicated.  I'd recommend moving this into the render body and giving it a name.  `showCertificationCard` is probably fine.

ojeytonwilliams commented on client/src/templates/Introduction/components/CertChallenge.js:
```diff
@@ -96,7 +125,6 @@ const CertChallenge = ({
       ).length + projectsCompleted;
     const numberOfSteps = Object.keys(steps).length;
-
     setCanViewCert(completedCount === numberOfSteps);
```
```suggestion
    setCanViewCert(completedCount === numberOfSteps);
```
I know it's out of scope, but `canViewCert` is misleading.  It's really just `hasCompletedRequiredSteps`.

Sembauke committed:
fix: variable names

ShaunSHamilton commented on client/src/templates/Introduction/components/CertChallenge.js:
```diff
@@ -90,8 +94,8 @@ const CertChallenge = ({
-          setCanClaim({ status, result });
-          setUserChecked(true);
+          setCertIsClaimable({ status, result });
+          SetVerificationFetchState(true);
```
```suggestion
          setVerificationFetchState(true);
```

ShaunSHamilton commented on client/src/templates/Introduction/components/CertChallenge.js:
```diff
@@ -73,15 +73,19 @@ const CertChallenge = ({
   const [isCertified, setIsCertified] = useState(false);
   const [userLoaded, setUserLoaded] = useState(false);
-  const [userChecked, setUserChecked] = useState(false);
+  const [verificationFetchState, SetVerificationFetchState] = useState(false);
```
```suggestion
  const [verificationFetchState, setVerificationFetchState] = useState(false);
```

Sembauke committed:
Apply suggestions from code review

Co-authored-by: Shaun Hamilton <shauhami020@gmail.com>

ojeytonwilliams committed:
fix: finish renaming variables

ojeytonwilliams reviewed (approved):
The feature itself works rather nicely.  It's a nice improvement, thanks @Sembauke :purple_heart: 

That is to say, LGTM :+1:

Sembauke commented:
No problem @ojeytonwilliams!

I agree that there are a lot of vague steps that have to be done in order to get this working properly.

But it's also a complicated process.
